<a href="https://colab.research.google.com/github/fralfaro/MAT281/blob/main/docs/labs/lab_03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# MAT281 - Laboratorio N°03





**Objetivo**: Aplicar técnicas avanzadas de manipulación y análisis de datos con pandas sobre un conjunto real de datos de contenido de Netflix, reforzando buenas prácticas y métodos eficientes sin recurrir a `groupby`, `merge`, `pivot`, ni `join`.



**Dataset**:

Trabajaremos con el archivo `netflix_titles.csv`, que contiene información sobre los títulos disponibles en la plataforma Netflix hasta el año 2021.

| Variable       | Clase     | Descripción                                                                 |
|----------------|-----------|------------------------------------------------------------------------------|
| show_id        | caracter  | Identificador único del título en el catálogo de Netflix.                   |
| type           | caracter  | Tipo de contenido: 'Movie' o 'TV Show'.                                     |
| title          | caracter  | Título del contenido.                                                       |
| director       | caracter  | Nombre del director (puede ser nulo).                                       |
| cast           | caracter  | Lista de actores principales (puede ser nulo).                              |
| country        | caracter  | País o países donde se produjo el contenido.                                |
| date_added     | fecha     | Fecha en la que el título fue agregado al catálogo de Netflix.              |
| release_year   | entero    | Año de lanzamiento original del título.                                     |
| rating         | caracter  | Clasificación por edad (por ejemplo: 'PG-13', 'TV-MA').                      |
| duration       | caracter  | Duración del contenido (minutos o número de temporadas para series).        |
| listed_in      | caracter  | Categorías o géneros en los que está clasificado el contenido.              |
| description    | caracter  | Breve sinopsis del contenido.                                               |




In [1]:
import pandas as pd

# Cargar datos
df = pd.read_csv('https://raw.githubusercontent.com/fralfaro/MAT281/main/docs/labs/data/netflix_titles.csv')
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...


### Parte 1: Limpieza y preparación

1. Revisar y describir el dataset:

   * ¿Cuántas filas y columnas tiene?
   * ¿Qué tipos de datos hay?
   * ¿Cuántos valores nulos hay por columna?

2. Transformar la columna `date_added` a tipo fecha.

3. Crear columnas auxiliares con `assign`:

   * Año (`year_added`)
   * Mes (`month_added`)



In [2]:
df.info()
df.dtypes
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

1. Por la información dada arriba, se ve que hay 8807 filas y 12 columnas, además los tipos de datos son object e int, y por último las columnas que si tienen valores son las siguientes (la tabla de arriba básicamebte): director 2634; cast 825; country 931; date_added 10; rating 4; duration 3.

In [4]:
import datetime
df['new_date'] = pd.to_datetime(df['date_added'].str.strip(), errors='coerce')
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description   new_date  
0  As her father nears the end of his life, filmm... 2021-09-25  
1  After crossing paths at a party, a Cape Town t... 2021-09-24  
2  To protect his family from a powerful drug lor... 2021-09-24  
3  Feuds, flirtations and toilet talk go down amo... 2021-09-24  
4  In a city of coaching centers known to train I... 2021-09-24

In [6]:
df = df.assign(
    year_added  = df['new_date'].dt.year.astype('Int64'),
    month_added = df['new_date'].dt.month.astype('Int64')
)
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description   new_date  year_added  \
0  As her father nears the end of his life, filmm... 2021-09-25        2021   
1  After crossing paths at a party, a Cape Town t... 2021-09-24        2021   
2  To protect his family from a powerful drug lor... 2021-09-24        2021   
3  Feuds, flirtations and toilet talk go down amo... 2021-09-24        2021   
4  In a city of coaching centers known to train I... 2021-09-24        2021   

   month_added  
0            9  
1            9  
2            9  
3            9  
4            9

## Parte 2: Técnicas avanzadas de pandas

4. Utilizar `.loc` para seleccionar películas (`type == 'Movie'`) que fueron agregadas después del año 2018.

5. Utilizar `str.contains()` y `str.extract()`:

   * Filtrar títulos que contienen la palabra 'love' (sin distinguir mayúsculas/minúsculas).
   * Extraer la duración en minutos para las películas desde la columna `duration`.

6. Aplicar `explode()` sobre la columna `listed_in` para obtener una fila por cada género.

7. Obtener un top 10 de géneros más frecuentes utilizando `value_counts()`.

8. Aplicar `where()` y `mask()` para marcar las películas de más de 120 minutos como contenido largo en una nueva columna.

9. Utilizar `.loc` para filtrar películas que cumplen con:

   * Más de 100 minutos de duración.
   * Rating igual a `'R'`.
   * País igual a `'United States'`.

10. Utilizar `.style` para formatear visualmente el top 10 de películas más largas.

In [10]:
df_4 = df.loc[(df['year_added'] == 2018) & (df['type'] == "Movie")]
df_4.head()

show_id   type                        title                     director  \
4229   s4230  Movie  Bill Hicks: One Night Stand             John Fortenberry   
4230   s4231  Movie      Bill Hicks: Reflections                          NaN   
4231   s4232  Movie       Bill Hicks: Relentless                  Chris Bould   
4232   s4233  Movie      Bill Hicks: Revelations                  Chris Bould   
4233   s4234  Movie         Bill Hicks: Sane Man  Kevin Booth, David Johndrow   

                         cast         country         date_added  \
4229               Bill Hicks  United Kingdom  December 31, 2018   
4230  Bill Hicks, Steve Hicks             NaN  December 31, 2018   
4231               Bill Hicks  United Kingdom  December 31, 2018   
4232               Bill Hicks  United Kingdom  December 31, 2018   
4233               Bill Hicks   United States  December 31, 2018   

      release_year rating duration        listed_in  \
4229          1991  TV-MA   28 min  Stand-Up Comedy   
4230          2015  TV-MA   34 min    Documentaries   
4231          1992  TV-MA   61 min  Stand-Up Comedy   
4232          1993  TV-MA   56 min  Stand-Up Comedy   
4233          1989  TV-MA   81 min  Stand-Up Comedy   

                                            description   new_date  \
4229  Bill Hicks remembers a wild night in 1989 and ... 2018-12-31   
4230  This documentary about comedian Bill Hicks off... 2018-12-31   
4231  In one of his most iconic performances, late c... 2018-12-31   
4232  In his final recorded special, the iconoclasti... 2018-12-31   
4233  Filmed in 1989 at a turning point in comedian ... 2018-12-31   

      year_added  month_added  
4229        2018           12  
4230        2018           12  
4231        2018           12  
4232        2018           12  
4233        2018           12

In [17]:
df_5a = df.loc[df['title'].str.contains("love", case=False, na=False)]
df_5a.head()

show_id     type                      title               director  \
25      s26  TV Show       Love on the Spectrum                    NaN   
158    s159    Movie    Love Don't Cost a Thing              Troy Byer   
159    s160    Movie             Love in a Puff         Pang Ho-cheung   
206    s207    Movie  LSD: Love, Sex Aur Dhokha       Dibakar Banerjee   
227    s228    Movie                Really Love  Angel Kristi Williams   

                                                  cast        country  \
25                                    Brooke Satchwell      Australia   
158  Nick Cannon, Christina Milian, Kenan Thompson,...  United States   
159  Miriam Chin Wah Yeung, Shawn Yue, Singh Hartih...      Hong Kong   
206  Nushrat Bharucha, Anshuman Jha, Neha Chauhan, ...          India   
227  Kofi Siriboe, Yootha Wong-Loi-Sing, Michael Ea...  United States   

             date_added  release_year rating   duration  \
25   September 21, 2021          2021  TV-14  2 Seasons   
158   September 1, 2021          2003  PG-13    101 min   
159   September 1, 2021          2010  TV-MA    103 min   
206     August 27, 2021          2010  TV-MA    112 min   
227     August 25, 2021          2020  TV-MA     95 min   

                                            listed_in  \
25     Docuseries, International TV Shows, Reality TV   
158                         Comedies, Romantic Movies   
159            Comedies, Dramas, International Movies   
206  Dramas, Independent Movies, International Movies   
227       Dramas, Independent Movies, Romantic Movies   

                                           description   new_date  year_added  \
25   Finding love can be hard for anyone. For young... 2021-09-21        2021   
158  A nerdy teen tries to make himself cool by ass... 2021-09-01        2021   
159  When the Hong Kong government enacts a ban on ... 2021-09-01        2021   
206  This provocative drama examines how the voyeur... 2021-08-27        2021   
227  A rising Black painter tries to break into a c... 2021-08-25        2021   

     month_added  
25             9  
158            9  
159            9  
206            8  
227            8

In [16]:
df_5b = df.loc[df['type'] == "Movie"].copy()
df_5b = df_5b.assign(duration_minutes = df_5b['duration'].str.extract(r'(\d+)').astype('Int64'))
df_5b[['title','duration','duration_minutes']].head()

title duration  duration_minutes
0               Dick Johnson Is Dead   90 min                90
6   My Little Pony: A New Generation   91 min                91
7                            Sankofa  125 min               125
9                       The Starling  104 min               104
12                      Je Suis Karl  127 min               127

In [19]:
df_6 = df.assign(
    listed_in = df['listed_in'].str.split(r',\s*')
).explode('listed_in')

df_6[['title', 'listed_in']].head()

title               listed_in
0  Dick Johnson Is Dead           Documentaries
1         Blood & Water  International TV Shows
1         Blood & Water               TV Dramas
1         Blood & Water            TV Mysteries
2             Ganglands          Crime TV Shows

In [25]:
df_7 = df.assign(
    listed_in = df['listed_in'].str.split(r',\s*')
).explode('listed_in')
df_7 = df_exploded['listed_in'].value_counts().head(10)
print(top_10_genres)


listed_in
International Movies        2752
Dramas                      2427
Comedies                    1674
International TV Shows      1351
Documentaries                869
Action & Adventure           859
TV Dramas                    763
Independent Movies           756
Children & Family Movies     641
Romantic Movies              616
Name: count, dtype: int64


In [28]:
df['duration_minutes'] = df['duration'].str.extract(r'(\d+)').astype('Int64')
df['length_category'] = df['duration_minutes'].astype('object')
df['length_category'] = df['length_category'].mask(df['duration_minutes'] > 120, 'Contenido largo')
df[['title', 'duration', 'duration_minutes', 'length_category']].head()

title   duration  duration_minutes length_category
0   Dick Johnson Is Dead     90 min                90              90
1          Blood & Water  2 Seasons                 2               2
2              Ganglands   1 Season                 1               1
3  Jailbirds New Orleans   1 Season                 1               1
4           Kota Factory  2 Seasons                 2               2

In [29]:
df_9 = df.loc[
    (df['duration_minutes'] > 100) &
    (df['rating'] == 'R') &
    (df['country'] == 'United States')
]

df_9.head()

show_id   type                        title               director  \
48      s49  Movie                 Training Day          Antoine Fuqua   
81      s82  Movie                         Kate  Cedric Nicolas-Troyan   
131    s132  Movie  Blade Runner: The Final Cut           Ridley Scott   
139    s140  Movie           Do the Right Thing              Spike Lee   
144    s145  Movie                  House Party        Reginald Hudlin   

                                                  cast        country  \
48   Denzel Washington, Ethan Hawke, Scott Glenn, T...  United States   
81   Mary Elizabeth Winstead, Jun Kunimura, Woody H...  United States   
131  Harrison Ford, Rutger Hauer, Sean Young, Edwar...  United States   
139  Danny Aiello, Ossie Davis, Ruby Dee, Richard E...  United States   
144  Christopher Reid, Christopher Martin, Robin Ha...  United States   

             date_added  release_year rating duration  \
48   September 16, 2021          2001      R  122 min   
81   September 10, 2021          2021      R  106 min   
131   September 1, 2021          1982      R  117 min   
139   September 1, 2021          1989      R  120 min   
144   September 1, 2021          1990      R  104 min   

                                           listed_in  \
48                                 Dramas, Thrillers   
81                                Action & Adventure   
131  Action & Adventure, Classic Movies, Cult Movies   
139                 Classic Movies, Comedies, Dramas   
144                            Comedies, Cult Movies   

                                           description   new_date  year_added  \
48   A rookie cop with one day to prove himself to ... 2021-09-16        2021   
81   Slipped a fatal poison on her final job, a rut... 2021-09-10        2021   
131  In a smog-choked dystopian Los Angeles, blade ... 2021-09-01        2021   
139  On a sweltering day in Brooklyn, simmering rac... 2021-09-01        2021   
144  Grounded by his strict father, Kid risks life ... 2021-09-01        2021   

     month_added  duration_minutes  length_category  
48             9               122  Contenido largo  
81             9               106              106  
131            9               117              117  
139            9               120              120  
144            9               104              104

In [30]:
df_10 = df.loc[df['type'] == 'Movie'].nlargest(10, 'duration_minutes')

# Formatear solo la columna de duración
df_10.style.format({'duration_minutes': '{:.0f} min'})




### Pregunta Desafío

11. ¿Cuáles son las combinaciones más frecuentes de género y rating en el dataset?
    (Sugerencia: utilizar `value_counts` con `subset=["genre", "rating"]` después de aplicar `explode()`).



### Bonus: Análisis de duplicados y limpieza

12. ¿Existen películas con el mismo nombre (`title`) pero con distinto año de lanzamiento (`release_year`)?
13. ¿Cuántos títulos únicos hay en total en la columna `title`?





In [31]:
df_exploded = df.assign(
    listed_in = df['listed_in'].str.split(r',\s*')
).explode('listed_in')
genre_rating_counts = df_exploded.value_counts(subset=['listed_in', 'rating']).head(10)
print(genre_rating_counts)


listed_in               rating
International Movies    TV-MA     1130
                        TV-14     1065
Dramas                  TV-MA      830
International TV Shows  TV-MA      714
Dramas                  TV-14      693
International TV Shows  TV-14      472
Comedies                TV-14      465
TV Dramas               TV-MA      434
Comedies                TV-MA      431
Dramas                  R          375
Name: count, dtype: int64
